server/prometheus.go

```diff
@@ -58,6 +74,43 @@ func (pp *PrometheusProvider) init() error {
 	return nil
 }
 
+func ExecuteGraphQuery(ctx *gin.Context, queryExpression string, env map[string][]string, duration time.Duration, pp *PrometheusProvider) (model.Value, v1.Warnings, error) {
+	tmpl, err := template.New("query").Parse(queryExpression)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	env1 := make(map[string]string)
+	for k, v := range env {
+		env1[k] = strings.Join(v, ",")
+	}
+
+	buf := new(bytes.Buffer)
+	err = tmpl.Execute(buf, env1)
+	if err != nil {
+		return nil, nil, err
```

Comment:
All returned errors should have proper message added. See example in my comment above.